crates/arroyo-df/src/rewriters.rs

```diff
@@ -426,6 +427,104 @@ impl TreeNodeRewriter for UnnestRewriter {
     }
 }
 
+pub struct AsyncUdfRewriter<'a> {
+    provider: &'a ArroyoSchemaProvider,
+}
+
+impl<'a> AsyncUdfRewriter<'a> {
+    pub fn new(provider: &'a ArroyoSchemaProvider) -> Self {
+        Self { provider }
+    }
+
+    fn split_async(
+        expr: Expr,
+        provider: &ArroyoSchemaProvider,
+    ) -> DFResult<(Expr, Option<(String, AsyncOptions, Vec<Expr>)>)> {
+        let mut c: Option<(String, AsyncOptions, Vec<Expr>)> = None;
+        let expr = expr.transform_up_mut(&mut |e| {
+            match &e {
+                Expr::ScalarFunction(ScalarFunction {
+                    func_def: ScalarFunctionDefinition::UDF(udf),
+                    args,
+                }) => {
+                    if let Some(UdfType::Async(opts)) =
+                        provider.udf_defs.get(udf.name()).map(|udf| udf.udf_type)
+                    {
+                        if c.replace((udf.name().to_string(), opts, args.clone()))
+                            .is_some()
+                        {
+                            return plan_err!(
+                                "multiple async calls in the same expression, which is not allowed"
+                            );
+                        }
+                        return Ok(Transformed::Yes(Expr::Column(Column::new_unqualified(
+                            ASYNC_RESULT_FIELD,
+                        ))));
+                    }
+                }
+                _ => {}
+            }
+            Ok(Transformed::No(e))
+        })?;
+
+        Ok((expr, c))
+    }
+}
+
+impl<'a> TreeNodeRewriter for AsyncUdfRewriter<'a> {
+    type N = LogicalPlan;
+
+    fn mutate(&mut self, node: Self::N) -> DFResult<Self::N> {
+        let LogicalPlan::Projection(mut projection) = node else {
+            for e in node.expressions() {
+                if let (_, Some((udf, _, _))) = Self::split_async(e.clone(), &self.provider)? {
+                    return plan_err!(
+                        "async UDFs are only supported in projections, but {udf} was called in another context"
+                    );
+                }
+            }
+            return Ok(node);
+        };
+
+        let mut args = None;
+
+        for e in projection.expr.iter_mut() {
+            let (new_e, Some(udf)) = Self::split_async(e.clone(), &self.provider)? else {
+                continue;
+            };
+
+            if args.replace(udf).is_some() {
+                return plan_err!(
+                    "projection {} contains multiple async UDFs, which is not supported",
```

Comment:
This prints out the whole logical plan up to this, which isn' t very user-understandable. e.g 
```
CREATE TABLE nexmark with (
    connector = 'nexmark',
    event_rate = '10'
);
SELECT new_udf(1), new_udf_two(1)  from nexmark where bid is not null  
```
gives a UI error like
```
Error during planning: projection Projection: __async_result, new_udf_two(Int64(1)), nexmark._timestamp
  Filter: nexmark.bid IS NOT NULL
    WaterMarkNode(nexmark): nexmark.person, nexmark.auction, nexmark.bid, nexmark._timestamp
      RemoteTableExtension: nexmark.person, nexmark.auction, nexmark.bid, nexmark._timestamp
        Projection: nexmark.person, nexmark.auction, nexmark.bid, nexmark._timestamp
          TableSourceExtension: nexmark.person, nexmark.auction, nexmark.bid, nexmark._timestamp contains multiple async UDFs, which is not supported
 ```

couple ways to fix this, but at least the "multiple async UDFs, which is not supported", should be at the beginning of the error.

Reply:
Changed it to

```
    Error during planning: Projection contains multiple async UDFs, which is not supported
    (hint: two async UDFs calls, get_city and new_udf, appear in the same SELECT statement)
```